### PR TITLE
Enable hiding/showing individual events in visualizations

### PR DESCRIPTION
**Pull request by gusaiani**
Solves #26869

### How to Test

Please check with visual and native questions.

a. Example of a visual question: `Sample Database › Orders, Summarize using Count by Created at: Month`
b. Example of a native query: `SELECT ID, CREATED_AT FROM ORDERS limit 1000`

Visualize question using Bars and Lines, among others

1. Click on the calendar icon on the bottom right
2. Add timeline events
3. Click on the checkbox next to each timeline event to hide/show it in the viz

#### Highlighted events

Events can be highlighted by clicking on their divs in the sidebar.

If hidden, atm an event cannot be highlighted.

If hidden when already highlighted, once made visible again, the event will be highlighted.

Highlighting a hidden event will no longer make a parent timeline visible in case it was hidden.

🎗️ I'm waiting on a look from @kdoh and @cdeweyx on the above.

Demo:
https://www.loom.com/share/5e2f5747e45b4ea0a54646c580f9d453

#### Timelines that are hidden on page load

On page load, timelines from collections other than the collection a question is in are hidden.
This is unchanged in this PR.
All child events of these hidden events will be shown as visible as the user makes their timelines visible.

Demo:
https://www.loom.com/share/724f1ad483be4d468220e7726cec64c3

### CODE REVIEW

File: frontend/src/metabase/timelines/questions/components/EventCard/EventCard.tsx [hunk 1]
```diff
@@ -56,6 +71,8 @@ const EventCard = ({
       isSelected={isSelected}
       onClick={handleEventClick}
     >
+      <Checkbox checked={!isHidden} onClick={handleChangeVisibility} />
```
Comment (gusaiani) [inline frontend/src/metabase/timelines/questions/components/EventCard/EventCard.tsx:74]: Ideally should be `onChange`, not `onClick`, but then the click propagates and toggles select. Which can be treated by reusing `handleAsideClick` but then it demands changing that function's name and things start getting polluted.

File: frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js [hunk 2]
```diff
@@ -256,6 +256,15 @@ describe("scenarios > organization > timelines > question", () => {
 
       cy.findByText("Visualization").should("be.visible");
       cy.findByLabelText("star icon").should("be.visible");
+
+      // should hide individual events from chart if hidden in sidebar
+      cy.icon("calendar").click();
```
Comment (ranquild) [inline frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:261]: This doesn't seem to have anything in common with the original intent of this test. Could we move it to a separate test?

**Review by ranquild — CHANGES_REQUESTED** [review]
The PR only tests that with a timeline with a single event it is possible to hide this event. I believe it's also worth checking:
- it works for multiple events, at least 2
- it is possible to make the event visible back
- it works correctly when combined with toggling the timeline on and off

Also it's not clear whether the behavior below is correct. Should we disable checkboxes and make them toggled off when the timeline is disabled?
Screenshot 2022-12-16 at 12 08 58

File: frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js (hunk 2 above)
Comment (gusaiani) [reply frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:261]: @ranquild We can, but at the cost of making e2e tests slower due to setup for each run.

@nemanjaglumac has been in favor of making longer individual e2e test cases.

Maybe we can add something to the title of the test to mention hiding?

File: frontend/src/metabase/query_builder/reducers.js [hunk 3]
```diff
@@ -558,6 +560,21 @@ export const visibleTimelineIds = handleActions(
   [],
 );
 
+export const hiddenTimelineEventIds = handleActions(
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/query_builder/reducers.js:563]: No fundamental block here, I think.

The consistency kept in this approach is quick "init" of reducer functions.

So when `handleActions`, in both `visibleTimelineIds` and `hiddenTimelineEventIds`, are initialized with `[INITIALIZE_QB]`, they will not process anything and just start with an empty array. 

That said, if it looks to you that we'd better make the naming consistent, yup, makes sense as well.

**Comment by gusaiani** [comment]
@ranquild Cool about the tests you mention, will add them.

> Also it's not clear whether the behavior below is correct. Should we disable checkboxes and make them toggled off when the timeline is disabled?

We can try that, I suppose provided the same hidden/shown events are restored once the timeline is made visible again.

File: frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js (hunk 2 above)
Comment (ranquild) [reply frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:261]: I'll let @nemanjaglumac decide what assertions should be combined in one test.

**Comment by nemanjaglumac** [comment]
Nit: There is a slight "jump" in the visualization itself when you toggle all elements off/on. Does it make sense to add the padding/margin between the X-axis and the graph to avoid this jump?

In other words, to anticipate the existance of event icons and to provide the space for them.

**Comment by gusaiani** [comment]
> Nit: There is a slight "jump" in the visualization itself when you toggle all elements off/on. Does it make sense to add the padding/margin between the X-axis and the graph to avoid this jump?

> In other words, to anticipate the existance of event icons and to provide the space for them.

Personally I don't mind the jump, as leaving the padding when there are no visible events (or full event timelines) will make the x-axis legends distant from the chart. 

I defer to our designers.

File: frontend/src/metabase/query_builder/reducers.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/query_builder/reducers.js]: Updated with `visible`

File: frontend/src/metabase/timelines/questions/components/EventCard/EventCard.tsx [hunk 4]
```diff
@@ -56,11 +70,14 @@ const EventCard = ({
       isSelected={isSelected}
       onClick={handleEventClick}
     >
-      <CardIconContainer>
-        <CardIcon name={event.icon} />
-      </CardIconContainer>
+      <CardCheckboxContainer data-testid="toggle-event-visibility">
+        <Checkbox checked={!isHidden} onClick={handleChangeVisibility} />
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/timelines/questions/components/EventCard/EventCard.tsx]: Awesome, done.

File: frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js (hunk 2 above)
Comment (gusaiani) [reply frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:277]: @ranquild @nemanjaglumac 

With the naming change from "hidden" to "visible" in all the props and functions, we have to do work to ensure a new event is visible upon creation.

We'd better test that, and we are.

Now the e2e test covers enough ground to warrant its own block, see if you disagree.

File: frontend/src/metabase/query_builder/actions/timelines.js [hunk 5]
```diff
@@ -14,6 +14,14 @@ export const selectTimelineEvents = createAction(SELECT_TIMELINE_EVENTS);
 export const DESELECT_TIMELINE_EVENTS = "metabase/qb/DESELECT_TIMELINE_EVENTS";
 export const deselectTimelineEvents = createAction(DESELECT_TIMELINE_EVENTS);
 
+export const HIDE_TIMELINE_EVENTS = "metabase/qb/HIDE_TIMELINE_EVENTS";
+export const hideTimelineEvents = createAction(HIDE_TIMELINE_EVENTS);
+
+export const SHOW_TIMELINE_EVENTS = "metabase/qb/SHOW_TIMELINE_EVENTS";
+export const showTimelineEvents = createAction(SHOW_TIMELINE_EVENTS);
+
+export const CREATE_TIMELINE_EVENT = "metabase/entities/timelineEvents/CREATE";
```
Comment (gusaiani) [inline frontend/src/metabase/query_builder/actions/timelines.js:23]: This is a default action type coming from entities.

I think we can use a ready-made constant for this, and I looked for an example, but so far I haven't found it.

Is there a ready-made?

File: frontend/src/metabase/query_builder/reducers.js [hunk 6]
```diff
@@ -558,6 +562,27 @@ export const visibleTimelineIds = handleActions(
   [],
 );
 
+export const visibleTimelineEventIds = handleActions(
+  {
+    [INITIALIZE_QB]: { next: () => [] },
+    [CREATE_TIMELINE_EVENT]: {
+      next: (state, { payload }) => [...state, payload.timelineEvent.id],
+    },
+    [INITIALIZE_VISIBLE_TIMELINE_EVENTS]: {
```
Comment (gusaiani) [inline frontend/src/metabase/query_builder/reducers.js:571]: @ranquild, here I add an action type just to handle initialization.

Two reasons:

1. to mark that this is about initializing, simply
2. show/hideTimelineEvent are about a single event. So all across the props and such this becomes tighter.

wdyt?

**Comment by gusaiani** [comment]
> If the intent was to mimic the MacOS behavior when you click the top-most checkbox, then all children should be either off or on.

@nemanjaglumac,
I've updated this.
Now the timeline events checkboxes appear unchecked when the timeline is hidden.
Once the timeline is shown again, the individual checkboxes that were shown/hidden before stay the same.

File: frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js (hunk 2 above)
Comment (gusaiani) [reply frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:277]: Tests have changed enough from first comment to warrant other conversations.

File: frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js [hunk 7]
```diff
@@ -257,6 +257,90 @@ describe("scenarios > organization > timelines > question", () => {
...
+        });
+
+      // making a hidden timeline visible
+      // should make its events automatically visible
+      cy.findByText("Timeline for collection")
+        .closest("[aria-label=Timeline card header]")
+        .within(() => cy.findByRole("checkbox").click());
+
+      cy.get(".x.axis").within(() => {
+        cy.findByLabelText("warning icon").should("be.visible");
```
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:333]: Nit: You can simply use `cy.icon("warning");`

Comment (gusaiani) [reply frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:333]: @nemanjaglumac 

> Nit: You can simply use cy.icon("warning").should("be.visible");

Thanks, updating.

Also, made this quick PR updating the calls in already-existing test cases in this file:
https://github.com/metabase/metabase/pull/27346

File: frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.unit.spec.tsx [hunk 8]
```diff
@@ -39,28 +39,14 @@ describe("TimelineCard", () => {
 
     expect(props.onToggleTimeline).toHaveBeenCalled();
   });
-
-  it("should make a timeline visible when its event is selected", () => {
```
Comment (gusaiani) [inline frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.unit.spec.tsx:43]: This test block is removed as long as we don't want to show a timeline by clicking on a hidden event in that timeline.

This is a design decision we're waiting on for now.

File: frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.unit.spec.tsx [hunk 9]
```diff
@@ -39,28 +39,14 @@ describe("TimelineCard", () => {
...
-    userEvent.click(screen.getByText("Releases"));
-    userEvent.click(screen.getByText("RC"));
-
-    expect(props.onToggleTimeline).toHaveBeenCalledWith(props.timeline, true);
-  });
 });
 
 const getProps = (opts?: Partial<TimelineCardProps>): TimelineCardProps => ({
   timeline: createMockTimeline(),
+  visibleEventIds: [],
```
Comment (gusaiani) [inline frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.unit.spec.tsx:46]: This and other addition to `getProps` functions in unit test files in this PR reflect added requirements in types used in the corresponding implementation files.

**Review by kdoh — CHANGES_REQUESTED** [review]
I think overall this is feeling pretty good with a couple small things to iron out still.

The parent / child relationship between timelines and events doesn't feel right as it exists in the branch right now, particularly the way that individual event visibility is disabled if the top level timeline is turned off. This breaks the ability to turn off the timeline to remove overall noise, and then selectively add back relevant events. There's [a description](https://www.notion.so/metabase/Allow-individual-events-to-be-toggled-on-off-eedd91cf28cc450f8106d51437f92a76#437b2f0d995748b88c54129303513a0e) in the product doc of how that's supposed to work. Happy to hop on a huddle or call if you need.

R.e. the jumping, it doesn't bother me a ton. It's really only an issue if you're turning everything off, in which case I think you want the extra space. We could look into a nicer transition, but that's relatively low on my list of priorities just this second.

File: frontend/src/metabase/query_builder/actions/timelines.js (hunk 5 above)
Comment (ranquild) [reply frontend/src/metabase/query_builder/actions/timelines.js:29]: `TimelineEvents.actionTypes.CREATE` or something like that

**Review by ranquild — CHANGES_REQUESTED** [review]
Please ping me when changes from the product doc mentioned by @kdoh are implemented.

**Comment by gusaiani** [comment]
@kdoh, thanks for your review.

PR is updated with the "dash" state of the checkbox.

I've made two variation PRs. Their descriptions should make them clear. Please let me know what you think

1. https://github.com/metabase/metabase/pull/27360
2. https://github.com/metabase/metabase/pull/27361

File: frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js [hunk 10]
```diff
@@ -257,6 +257,90 @@ describe("scenarios > organization > timelines > question", () => {
...
+        timeline: { name: "Timeline for collection", collection_id: 1 },
+        events: [
+          { name: "TC1", timestamp: "2016-05-20T00:00:00Z", icon: "warning" },
+        ],
+      });
+
+      visitQuestion(3);
+
+      cy.findByText("Visualization").should("be.visible");
+      cy.findByLabelText("cloud icon").should("be.visible");
```
Comment (gusaiani) [inline frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:279]: @nemanjaglumac,

I've checked and we do need to use `findByLabelText` to select these icons in the chart.

These event icons are rendered as svg paths, and the class names we use to select with `cy.icon` are not added to them.

File: frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js (hunk 7 above)
Comment (gusaiani) [reply frontend/test/metabase/scenarios/organization/timelines-question.cy.spec.js:333]: @nemanjaglumac,

I've checked and we do need to use findByLabelText to select these icons in the chart.

These event icons are rendered as svg paths, and the class names we use to select with cy.icon are not added to them.

File: frontend/src/metabase/query_builder/actions/timelines.js (hunk 5 above)
Comment (gusaiani) [reply frontend/src/metabase/query_builder/actions/timelines.js:29]: Yup, that. Thank you

**Review by kdoh — APPROVED** [review]
Event and timeline select / de-select behavior seems to work as I expect now 👍 Nice work!

**Review by nemanjaglumac — APPROVED** [review]
Overall, E2E test looks and works much better.
I tested the behaviour manually and can confirm that it also makes more sense now after a redesign.

**Review by ranquild — CHANGES_REQUESTED** [review]
I believe we shouldn't keep track of timeline visibility when we track visibility of each event.
In my opinion, the timeline visibility is a property that should be computed based on visibility of its events. Could you remove `visibleTimelineIds` reducer and implement the changes above?

Also, in this PR hovering over a visible event doesn't make it highlighted in the query builder. Something got broken.

Another thing is that unit testing reducers is an anti-pattern https://redux.js.org/usage/writing-tests. Could you test the connected component instead (with `renderWithProviders`)?

**Comment by gusaiani** [comment]
@ranquild Yup to all, on it, thanks.

**Comment by gusaiani** [comment]
@ranquild 

You mention not unit-testing reducers, and perhaps using `renderWithProviders` to the same effect.

I see the idea behind not testing the reducers, and explored using `renderWithProviders` there.
The thing is, the depth of components being passed props is quite significant.
It didn't seem that adding test to the ancestor component, basically the entire visualization, would be practical.
We could inject selectors down the component pipeline, which is not the way we've been doing things.
Considering we cover the cases anyway in the e2e tests, wdyt of us simply not having the reducer tests in this situation?

File: frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.tsx [hunk 11]
```diff
@@ -23,30 +24,40 @@ import {
...
   onToggleEventSelected,
-  onToggleTimeline,
+  onShowTimelineEvents,
+  onHideTimelineEvents,
 }: TimelineCardProps): JSX.Element => {
   const events = getEvents(timeline.events);
   const isEventSelected = events.some(e => selectedEventIds.includes(e.id));
   const [isExpanded, setIsExpanded] = useState(isDefault || isEventSelected);
 
+  const anyEventVisible = events.some(event =>
```
Comment (ranquild) [inline frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.tsx:53]: nit: maybe use `useMemo`? seems a bit time consuming

**Review by ranquild — APPROVED** [review]
Only a comment about useMemo

File: frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.tsx (hunk 11 above)
Comment (ranquild) [reply frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.tsx:53]: And below as well

Comment (gusaiani) [reply frontend/src/metabase/timelines/questions/components/TimelineCard/TimelineCard.tsx]: Good one, added.

**Pull request merged**